### PR TITLE
Measurements_from_sample transform

### PR DESCRIPTION
**Pull request by lillian542**
**Context:**
Catalyst already includes a transform to convert all measurement processes into counts + post-processing, but doesn't have a corresponding transform for devices that return samples.

**Description of the Change:**
- Adds a transform that decomposes to `sample` rather than `counts`
- Small adjustment to `measurements_to_counts` to support measurements with no wires (`qml.probs()`, `qml.counts()`)

### CODE REVIEW

**Comment by lillian542** [comment]
[sc-60183]

**Comment by lillian542** [comment]
The final CodeFactor issue is complaining about the test file being too long - seems like a reasonable complaint, but I didn't want to make the diff confusing in terms of which tests are being changed or added vs just moved, so I have that in a separate PR [here](https://github.com/PennyLaneAI/catalyst/pull/1116)

File: frontend/catalyst/device/decomposition.py [hunk 1]
```diff
@@ -223,56 +226,125 @@ def measurements_from_counts(tape):
...
+        measured_wires (list): A list of all wires that are measured on the tape
+
+    """
+    if tape.samples_computational_basis and len(tape.measurements) > 1:
+        _validate_computational_basis_sampling(tape)
+
+    diagonalizing_gates, diagonal_measurements = rotations_and_diagonal_measurements(tape)
+    for m in diagonal_measurements:
+        if m.obs is not None:
+            diagonalizing_gates.extend(m.obs.diagonalizing_gates())
```
[the remark this answers is not in the document]
Comment (lillian542) [reply frontend/catalyst/device/decomposition.py:334]: Hahaha, yes, you would think `rotations_and_diagonal_measurements` rotated everything, but it actually only rotates _overlapping_ measurements (i.e. if X(0) appears more than one place, it diagonalizes that). I think it's one of the functions in PennyLane with the very least informative name. 

As an example, here it only diagonalizes `X(0)`, not `Y(1)`:

```
>>> from pennylane.tape.tape import rotations_and_diagonal_measurements

>>> measurements=[qml.expval(qml.X(0)), qml.var(qml.X(0)), qml.expval(qml.Y(1))]
>>> tape = qml.tape.QuantumScript([], measurements)

>>> rotations_and_diagonal_measurements(tape)
([RY(-1.5707963267948966, wires=[0])], [expval(Z(0)), var(Z(0)), expval(Y(1))])
```

File: frontend/catalyst/device/decomposition.py [hunk 2]
```diff
@@ -223,56 +226,125 @@ def measurements_from_counts(tape):
...
+        if len(tape.measurements) == 1:
+            results_processed = results_processed[0]
+        else:
+            results_processed = tuple(results_processed)
+        return results_processed
+
+    return [new_tape], postprocessing_samples
+
+
+def _diagonalize_measurements(tape, device_wires):
```
[the remark this answers is not in the document]
Comment (lillian542) [reply frontend/catalyst/device/decomposition.py:316]: It kind of is, maybe it's silly to have this be separate. Getting the measurement wires is not part of the existing diagonalization functions, and they do "extra work" in creating diagonalized observables that we won't need anyway because we are taking samples directly with no observables in this transform. But that's not a lot of extra work, and we could combine them to have fewer things-that-do-almost-exactly-the-same-thing 🤷‍♀️ Maybe we should?

File: frontend/catalyst/device/decomposition.py [hunk 3]
```diff
@@ -282,17 +354,17 @@ def _get_probs(counts_outcome):
...
     expval = jax.numpy.dot(jax.numpy.array(eigvals), prob_vector)
     return expval
 
 
-def _get_var(eigvals, prob_vector):
+def _var_from_probs(eigvals, prob_vector):
     """From the observable eigenvalues and the probability vector
     it calculates the variance."""
-    var = jax.numpy.dot(prob_vector, (eigvals**2)) - jax.numpy.dot(prob_vector, eigvals)
+    var = jax.numpy.dot(prob_vector, (eigvals**2)) - jax.numpy.dot(prob_vector, eigvals) ** 2
```
[the remark this answers is not in the document]
Comment (lillian542) [reply frontend/catalyst/device/decomposition.py:367]: This slipped under the radar because we were testing that things could run, and produced the right type and shape of result, but not the result was correct compared to a trusted device or an analytic result 🐛

File: frontend/catalyst/device/qjit_device.py [hunk 4]
```diff
@@ -519,8 +520,10 @@ def _measurement_transform_program(self):
         elif not supports_sum_observables:
             measurement_program.add_transform(split_to_single_terms)
 
+        if self.measurement_processes in [{"Sample"}, {"Counts", "Sample"}]:
+            measurement_program.add_transform(measurements_from_samples, self.wires)
         if self.measurement_processes == {"Counts"}:
-            measurement_program.add_transform(measurements_from_counts)
+            measurement_program.add_transform(measurements_from_counts, self.wires)
```
[the remark this answers is not in the document]
Comment (lillian542) [reply frontend/catalyst/device/qjit_device.py:526]: Yes, exactly. I don't think there's a clear advantage of one over the other, but we have to pick, and samples feels more "default"

File: frontend/test/pytest/test_transform.py [hunk 5]
```diff
@@ -365,23 +366,20 @@ def qfunc():
...
 
-            @sum_expand
+            @qml.transforms.split_non_commuting
             @qml.qnode(qml.device(device_name, wires=2, shots=None))
             def qfunc():
                 """Example taken from PL tests"""
                 obs1 = qml.prod(qml.PauliX(0), qml.PauliX(1))
                 obs2 = qml.prod(qml.PauliX(0), qml.PauliY(1))
-                return [qml.expval(obs1), qml.expval(obs2)]
+                return qml.expval(obs1), qml.expval(obs2)
```
[the remark this answers is not in the document]
Comment (lillian542) [reply frontend/test/pytest/test_transform.py:382]: I would say cosmetic? 

With the list, the transform runs fine, and the returned values still matched expectation, but `assert expected_shape == observed_shape` failed because the `observed_shape` returned a list, and the `expected_shape` a tuple, so the PyTrees weren't the same.

File: frontend/catalyst/device/decomposition.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/device/decomposition.py:334]: Got it, thanks 😅 Maybe a comment would be useful for future readers

File: frontend/catalyst/device/decomposition.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dime10) [reply frontend/catalyst/device/decomposition.py:316]: Hmm I see, we'll I think it's fine as is as well :) Of course if you want to streamline in a follow up or in the future that's cool too 😄

File: frontend/catalyst/device/decomposition.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lillian542) [reply frontend/catalyst/device/decomposition.py]: I just replaced this part with the `diagonalize_measurements` transform from PL, it removes the confusion around this and avoids creating duplicate functionality.
